src/one/helfy/vmstruct/Frame.java

```diff
@@ -0,0 +1,212 @@
+    private static final int slot_sender_sp = 2;
+
+    private final long sp;
+    private final long unextendedSP;
+    private final long fp;
+    private final long pc;
+
+    public Frame(long sp, long fp, long pc) {
+        this(sp, sp, fp, pc);
+    }
+
+    public Frame(long sp, long unextendedSP, long fp, long pc) {
+        this.sp = sp;
+        this.unextendedSP = unextendedSP;
+        this.fp = fp;
+        this.pc = pc;
+    }
+
+    public long at(int slot) {
+        return jvm.getAddress(fp + slot * wordSize);
+    }
+
+    public long local(int index) {
+        return jvm.getAddress(at(slot_interp_locals) - index * wordSize);
+    }
+
+    public int bci() {
+        long bcp = at(slot_interp_bcp);
+        if (0 <= bcp && bcp <= 0xffff) {
+            return (int) bcp;
+        }
+        return Method.bcpToBci(at(slot_interp_method), bcp);
+    }
+
+    public long method() {
+        if (Interpreter.contains(pc)) {
+            return at(slot_interp_method);
+        }
+
+        long cb = CodeCache.findBlob(pc);
+        if (cb != 0) {
+            String name = jvm.getStringRef(cb + _name);
+            if (name.endsWith("nmethod")) {
+                return jvm.getAddress(cb + _method);
+            }
+        }
+
+        return 0;
+    }
+
+    public Frame sender() {
+        if (Interpreter.contains(pc)) {
+            return new Frame(fp + slot_sender_sp * wordSize, at(slot_interp_sender_sp), at(slot_link), at(slot_return_addr));
+        }
+
+        long cb = CodeCache.findBlob(pc);
+        if (cb != 0) {
+            long senderSP = unextendedSP + jvm.getInt(cb + _frame_size);
+            if (senderSP != sp) {
+                long senderPC = jvm.getAddress(senderSP - slot_return_addr * wordSize);
+                long savedFP = jvm.getAddress(senderSP - slot_sender_sp * wordSize);
+                return new Frame(senderSP, savedFP, senderPC);
+            }
+        }
+
+        return null;
+    }
+
+    private String valueAsText(long value) {
+        if (value >= ' ' && value <= '~') {
+            return value + " '" + (char) value + "'";
+        } else {
+            return Long.toString(value);
+        }
+    }
+
+    public void dumpLocals(PrintStream out, long method) {
+        int maxLocals = Method.maxLocals(method);
+        Method.LocalVar[] vars = Method.getLocalVars(method);
+        for (int i = 0; i < maxLocals; i++) {
+            long localVarVal = local(i);
+            boolean skipNext = false;
+            String valAsText = null;
+            if (vars.length > 0 && vars[i] != null) {
+                if (!vars[i].valueType && !vars[i].isArray) {
+                    ObjRef strRef = new ObjRef();
+                    // if we have -XX:-UseCompressedOops specified base and offset will be 0
+                    strRef.ptr = (int) ((localVarVal - _narrow_oop_base) >> _narrow_oop_shift);
+                    Object val = jvm.getObject(strRef, jvm.fieldOffset(ObjRef.ptrField));
```

Comment:
This operation is inherently unsafe, since GC may happen between `strRef.ptr` and `jvm.getObject`.